stdlib/public/core/UnsafeUninitializedPointer.swift

```diff
@@ -0,0 +1,140 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2021 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+/// Provides scoped access to a raw buffer pointer with the specified byte count
+/// and alignment.
+///
+/// - Parameters:
+///   - byteCount: The number of bytes to temporarily allocate. `byteCount` must
+///     not be negative.
+///   - alignment: The alignment of the new, temporary region of allocated
+///     memory, in bytes.
+///   - body: A closure to invoke and to which the allocated buffer pointer
+///     should be passed.
+///
+///  - Returns: Whatever is returned by `body`.
+///
+///  - Throws: Whatever is thrown by `body`.
+///
+/// This function is useful for cheaply allocating raw storage for a brief
+/// duration. Storage may be allocated on the heap or on the stack, depending on
+/// the required size and alignment.
+///
+/// When `body` is called, the contents of the buffer pointer passed to it are
+/// in an unspecified, uninitialized state. `body` is responsible for
+/// initializing the buffer pointer before it is used _and_ for deinitializing
+/// it before returning. `body` does not need to deallocate the buffer pointer.
+///
+/// The implementation may allocate a larger buffer pointer than is strictly
+/// necessary to contain `byteCount` bytes. The behavior of a program that
+/// attempts to access any such additional storage is undefined.
+///
+/// The buffer pointer passed to `body` (as well as any pointers to elements in
+/// the buffer) must not escape—it will be deallocated when `body` returns and
+/// cannot be used afterward.
+@_transparent
```

Comment:
`@_transparent` still generates an ABI symbol, so as it stands these functions are lacking an `@available` declaration. We also generally prefer `@inline(__always)` in favor of `@_transparent` for new additions, unless there is a technical reason why the inlining needs to happen early on. (Is there?)

Given that the implementations are (iiuc) back-deployable, I recommend replacing `@_transparent` with:

```suggestion
@_alwaysEmitIntoClient
@inline(__always)
```

This gets rid of the superfluous ABI symbol, and still inlines these calls into clients.

(This applies to all three of the new public functions introduced here.)

Reply:
The proposal under review specifically states that this API will be `@_transparent`.

Reply:
`@_transparent` is a non-public implementation detail that has never gone through the S-E process; I don't think the proposal should dictate such things. @grynspan I think it would make sense to remove this attribute from the API proposal.

Reply:
@lorentey We can replace it (I don't think we'll need to re-run the review for that.)